### DISCUSSION TITLE
Model Assembly for more than one meshes

### DISCUSSION THREAD

**Discussion by kmahajan-cadfem** [ansys/pyprimemesh#915] [Ideas]
There are many customers who leverage the model assembly where they can mesh smaller parts of the bigger model and bring them together using the Workbench coupling system.

Is there something like this that already exists in PyPrimeMeshing?

If not, I think this would be one of the most awaited feature, and with power of Python, customers can create a very powerful lightweight workflows for different situations.

**Comment by kmahajan-cadfem** [comment]
@waltersma any thoughts on above, is this supported or planned?

**Comment by waltersma** [comment]
Hi @kmahajan-cadfem,  

It is possible to read and amend various files of different types, mesh, copy, apply transformations and export solver files in PyPrimeMesh.  Do these operations give you what you need?   It might be useful to provide more details on a specific use case and highlight if there are specific gaps to what you require.  I hope these comments help.

Thanks

**Reply by kmahajan-cadfem** [reply]
Hi
thanks for the comment, could you maybe point me to an example or at least tests from which we can maybe deduce how to use all the above functionally from prime, I am mostly interested in copying/importing mesh file and then placing it in global coordinate systems on specific location and then write it out in cdb and also fluent meshes. 

Thanks

**Comment by waltersma** [comment]
Hi @kmahajan-cadfem,

Please see links below to apis related to the tasks you specified:

[get_cell_zonelets](https://prime.docs.pyansys.com/version/stable/api/_autosummary/ansys.meshing.prime.Part.get_cell_zonelets.html#ansys.meshing.prime.Part.get_cell_zonelets)
[copy_cell_zonelets](https://prime.docs.pyansys.com/version/stable/api/_autosummary/ansys.meshing.prime.VolumeMeshTool.copy_cell_zonelets.html#copy-cell-zonelets)
[transform_zonelets](https://prime.docs.pyansys.com/version/stable/api/_autosummary/ansys.meshing.prime.Transform.transform_zonelets.html#transform-zonelets)

An example code to show bringing these items together to copy a block of hex mesh read from a cdb file, move and connect before exporting as a Fluent mesh file:

```
import ansys.meshing.prime as prime

prime_client = prime.launch_prime()
model = prime_client.model
mesh_util = prime.lucid.Mesh(model=model)

# retrieve a cdb used for unit testing of a hex meshed block
download_manager = prime.examples.download_utilities.DownloadManager()
file_name=download_manager.download_file("hex.cdb", 'pyprimemesh', 'unit_test_examples')
mesh_util.read(file_name=file_name)

part = model.parts[0]
face_zonelets = part.get_face_zonelets()
cell_zonelets = part.get_cell_zonelets()

# copy the hex mesh as a new zone
copy_params = prime.CopyZoneletsParams(model, copy_zones=False)
copy_result = prime.VolumeMeshTool(model).copy_cell_zonelets(cell_zonelets, part.id, copy_params)
copied_cell_zonelets = copy_result.copied_zonelets
copied_face_zonelets = copy_result.copied_face_zonelets

# transform copy in x-direction to be adjacent to original
transform = [
    1, 0, 0, 0.032,
    0, 1, 0, 0,
    0, 0, 1, 0,
    0, 0, 0, 1, 
]
transform_params = prime.TransformParams(model, transform)
prime.Transform(model).transform_zonelets(part.id, copied_cell_zonelets, transform_params)

# use a beta api to merge nodes and connect the two meshes
merge_params = prime.MergeBoundaryNodesParams(
    model,
    merge_node_type=prime.MergeNodeType.ALLALL,
    tolerance=0.0001,
    use_absolute_tolerance=True,
)
result = prime.Connect(model).merge_boundary_nodes(part.id, face_zonelets, copied_face_zonelets, merge_params)

# write as a fluent mesh file
mesh_util.write(file_name="hex.msh")

prime_client.exit()
```

I hope this gives you what you need!

**Reply by susannehaase** [reply]
Hi @waltersma, 
if I run the example from above with a matrix for mirroring, I get negative volumes. Here is my code part:

```
#...
# transform copy in x-direction to be adjacent to original
transform = [
    1, 0, 0, 0.032,
    0, -1, 0, 0,
    0, 0, 1, 0,
    0, 0, 0, 1, 
]
transform_params = prime.TransformParams(model, transform)
prime.Transform(model).transform_zonelets(part.id, copied_cell_zonelets, transform_params)

# use a beta api to merge nodes and connect the two meshes
merge_params = prime.MergeBoundaryNodesParams(
    model,
    merge_node_type=prime.MergeNodeType.ALLALL,
    tolerance=0.0001,
    use_absolute_tolerance=True,
)
result = prime.Connect(model).merge_boundary_nodes(part.id, face_zonelets, copied_face_zonelets, merge_params)

# write as a fluent mesh file
mesh_util.write(file_name="hex.msh")

vtool = prime.VolumeMeshTool(model=model)
result = vtool.check_mesh(part_id=part.id, params=prime.CheckMeshParams(model=model))

print("Non positive volumes:", result.has_non_positive_volumes)
print("Non positive areas:", result.has_non_positive_areas)
print("Invalid shape:", result.has_invalid_shape)
print("Left handed faces:", result.has_left_handed_faces)
```
Output:
```
Non positive volumes: True
Non positive areas: False
Invalid shape: False
Left handed faces: True
```

This is an issue, right? I used the version 0.7.0.

Best regards, Susanne

**Comment by waltersma** [comment]
Added as an issue #1007
